subxt/src/backend/utils.rs

```diff
@@ -0,0 +1,244 @@
+    Stream(StreamOfResults<T>),
+}
+
+/// Retry subscription.
+struct RetrySubscription<T> {
+    resubscribe: ResubscribeGetter<T>,
+    state: PendingOrStream<T>,
+}
+
+impl<T> std::marker::Unpin for RetrySubscription<T> {}
+
+impl<T> Stream for RetrySubscription<T> {
+    type Item = Result<T, Error>;
+
+    fn poll_next(
+        mut self: Pin<&mut Self>,
+        cx: &mut std::task::Context<'_>,
+    ) -> Poll<Option<Self::Item>> {
+        enum NextState<T> {
+            Stream(StreamOfResults<T>),
+            Resubscribe(BoxFuture<'static, Result<StreamOfResults<T>, Error>>),
+        }
+
+        loop {
+            let next_state = match self.state {
+                PendingOrStream::Stream(ref mut s) => match s.poll_next_unpin(cx) {
+                    Poll::Ready(Some(Err(e))) => {
+                        if e.is_disconnected_will_reconnect() {
+                            NextState::Resubscribe((self.resubscribe)())
+                        } else {
+                            return Poll::Ready(Some(Err(e)));
+                        }
+                    }
+                    other => return other,
+                },
+                PendingOrStream::Pending(ref mut fut) => match ready!(fut.poll_unpin(cx)) {
+                    Ok(stream) => NextState::Stream(stream),
+                    Err(e) => {
+                        if !e.is_disconnected_will_reconnect() {
+                            return Poll::Ready(None);
+                        }
+
+                        NextState::Resubscribe((self.resubscribe)())
+                    }
+                },
+            };
+
+            match next_state {
+                NextState::Resubscribe(fut) => {
+                    self.state = PendingOrStream::Pending(fut);
+                }
+                NextState::Stream(stream) => {
+                    self.state = PendingOrStream::Stream(stream);
+                }
+            }
+        }
+    }
+}
+
+/// Retry a future until it doesn't return a disconnected error.
+pub async fn retry<T, F, R>(mut retry_future: F) -> Result<R, Error>
+where
+    F: FnMut() -> T,
+    T: Future<Output = Result<R, Error>>,
+{
+    const REJECTED_MAX_RETRIES: usize = 10;
+    let mut rejected_retries = 0;
+
+    loop {
+        match retry_future().await {
+            Ok(v) => return Ok(v),
+            Err(e) => {
+                if e.is_disconnected_will_reconnect() {
+                    continue;
+                }
+
+                // This applies only to the rpc v2 method calls/unstable backend.
+                //
+                // Once the backend sees `Error::DisconnectedWillReconnect` it will
+                // update the state, and subsequent `subscribe` will not proceed until
+                // a new subscription ID is obtained.
+                //
+                // It's still possible that subscription ID could be read before
+                // the backend has updated the state, but it is quite unlikely.
```

Comment:
I guess this depends on the ordering that futures/streams are polled in.

Eg take this backend function:

```rust
async fn block_header(&self, at: T::Hash) -> Result<Option<T::Header>, Error> {
    retry(|| async {
        let sub_id = get_subscription_id(&self.follow_handle).await?;
        self.methods.chainhead_v1_header(&sub_id, at).await
    })
    .await
}
```

Where `get_subscription_id` is basically `follow_handle.subscribe().subscription_id().await`.

If `self.methods.chainhead_v1_header(&sub_id, at).await` returns `DisconnectedWillReconnect`, we immediately re-run this async block which calls `get_subscription_id(&self.follow_handle).await?`, but if the `FollowStreamDriver` hasn't noticed that there's an error yet because it hasn't been polled enough to see a `Stop` event, then you'll get back a stale subscription ID from it (pretty much synchronously, because it's cached).

One solution might be to keep track of the subscription ID we are using. If we get given the same ID again after seeing a `DisconnectedWillReconnect` error, then we know it's wrong now and we need to try again. That could be done hackily in eg the code above, but I think there's probably a nicer way to guarantee that we get back a valid subscription ID while keeping the above clean.

I think it's worth raising an issue to think about how best to fix this, but for now I'm cool with this hack :)